editor/store/selectors.js

```diff
@@ -1202,152 +1202,280 @@ export function getNotices( state ) {
+ * inserter and handle its selection.
+ *
+ * The 'utility' property indicates how useful we think an item will be to the
+ * user. There are 4 levels of utility:
+ *
+ * 1. Blocks that are contextually useful (utility = 3)
+ * 2. Blocks that have been previously inserted (utility = 2)
+ * 3. Blocks that are in the common category (utility = 1)
+ * 4. All other blocks (utility = 0)
+ *
+ * The 'frecency' property is a herustic (https://en.wikipedia.org/wiki/Frecency)
+ * that combines block usage frequenty and recency.
+ *
+ * Items are returned ordered descendingly by their 'utility' and 'frecency'.
+ *
+ * @param {Object} state     Global application state.
+ * @param {?string} parentUID The block we are inserting into, if any.
  *
  * @return {Editor.InserterItem[]} Items that appear in inserter.
- */
-export function getInserterItems( state, allowedBlockTypes ) {
-	if ( ! allowedBlockTypes ) {
-		return [];
-	}
+ *
+ * @typedef {Object} Editor.InserterItem
+ * @property {string}   id                Unique identifier for the item.
+ * @property {string}   name              The type of block to create.
+ * @property {Object}   initialAttributes Attributes to pass to the newly created block.
+ * @property {string}   title             Title of the item, as it appears in the inserter.
+ * @property {string}   icon              Dashicon for the item, as it appears in the inserter.
+ * @property {string}   category          Block category that the item is associated with.
+ * @property {string[]} keywords          Keywords that can be searched to find this item.
+ * @property {boolean}  isDisabled        Whether or not the user should be prevented from inserting
+ *                                        this item.
+ * @property {number}   utility           How useful we think this item is, between 0 and 3.
+ * @property {number}   frecency          Hueristic that combines recency and frecency.
+ */
+export const getInserterItems = createSelector(
+	( state, parentUID = null ) => {
+		/**
+		 * The second argument used to be {boolean|array} allowedBlockTypes but it is no
+		 * longer necessary.
+		 */
+		if ( isBoolean( parentUID ) || isArray( parentUID ) ) {
+			deprecated( 'allowedBlockTypes', {
+				version: '3.1',
+				plugin: 'Gutenberg',
+			} );
+			parentUID = null;
+		}
 
-	const staticItems = getBlockTypes().map( ( blockType ) =>
-		buildInserterItemFromBlockType( state, allowedBlockTypes, blockType )
-	);
+		const calculateUtility = ( category, count, isContextual ) => {
+			if ( isContextual ) {
+				return 3;
+			} else if ( count > 0 ) {
+				return 2;
+			} else if ( category === 'common' ) {
+				return 1;
+			}
+			return 0;
+		};
 
-	const dynamicItems = getSharedBlocks( state ).map( ( sharedBlock ) =>
-		buildInserterItemFromSharedBlock( state, allowedBlockTypes, sharedBlock )
-	);
+		const calculateFrecency = ( time, count ) => {
+			if ( ! time ) {
+				return count;
+			}
+
+			const duration = Date.now() - time;
```

Comment:
I think it's fine since _"Now" in this case is the last time the relevant state changed_ is exactly what we want.

Using the greatest time from all of the blocks is a good idea for a pure alternative, but I don't think it's worth the performance penalty we'd get from iterating through every block.

Reply:
>I think it's fine since "Now" in this case is the last time the relevant state changed is exactly what we want.

This would have made for a good inline code comment to the next person who maintains this selector.